File: bot.js
// Polymarket Top Trader Tracker — Telegram Bot
// Run with: node bot.js

const BASE = "https://data-api.polymarket.com";
const POLYMARKET_URL = "https://polymarket.com/event";

const TELEGRAM_TOKEN  = process.env.TELEGRAM_TOKEN  || "";
const TELEGRAM_CHAT   = process.env.TELEGRAM_CHAT   || "";
const POLL_INTERVAL   = parseInt(process.env.POLL_INTERVAL || "120") * 1000;
const PERIOD          = process.env.LEADERBOARD_PERIOD || "ALL";
const MIN_SIZE        = parseFloat(process.env.MIN_TRADE_SIZE || "0.50");

// swisstony is always tracked — active trader with strong history
const FIXED_WALLET = {
  wallet: "0x204f72f35326db932158cba6adff0b9a1da95e14",
  name:   "swisstony",
  rank:   1,
  pnl:    0,
  fixed:  true,
};

const seenTx = new Set();
let trackedWallets = [];
let lastLeaderboardRefresh = 0;
const LEADERBOARD_TTL = 3600 * 1000;

// ── Helpers ──────────────────────────────────────────────────────────────────
function fmtMoney(n) {
  n = parseFloat(n || 0);
  const sign = n >= 0 ? "+" : "-";
  const abs = Math.abs(n);
  if (abs >= 1e6) return `${sign}$${(abs/1e6).toFixed(2)}M`;
  if (abs >= 1e3) return `${sign}$${(abs/1e3).toFixed(1)}K`;
  return `${sign}$${abs.toFixed(2)}`;
}

function fmtSize(n) {
  n = parseFloat(n || 0);
  if (n >= 1e6) return `$${(n/1e6).toFixed(2)}M`;
  if (n >= 1e3) return `$${(n/1e3).toFixed(1)}K`;
  return `$${n.toFixed(2)}`;
}

function rankEmoji(r) { return ["🥇","🥈","🥉"][r-1] || `#${r}`; }

function fmtTime(ts) {
  const d = new Date(ts * 1000);
  return d.toUTCString().replace(" GMT", " UTC").slice(5, 22);
}

function nowStr() { return new Date().toISOString().slice(11,19); }

// ── Fetch helpers (no dependencies, uses built-in fetch) ──────────────────────
async function get(url) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HTTP ${res.status} — ${url}`);
  return res.json();
}

async function fetchLeaderboard() {
  return get(`${BASE}/v1/leaderboard?limit=3&orderBy=PNL&timePeriod=${PERIOD}`);
}

async function fetchTrades(wallet) {
  const data = await get(`${BASE}/trades?user=${wallet}&limit=20&takerOnly=false`);
  return Array.isArray(data) ? data : [];
}

// ── Telegram ──────────────────────────────────────────────────────────────────
async function sendTelegram(text) {
  if (!TELEGRAM_TOKEN || !TELEGRAM_CHAT) {
    console.log("\n── ALERT (no Telegram configured) ──");
    console.log(text.replace(/<[^>]+>/g, ""));
    return;
  }
  const res = await fetch(`https://api.telegram.org/bot${TELEGRAM_TOKEN}/sendMessage`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      chat_id: TELEGRAM_CHAT,
      text,
      parse_mode: "HTML",
      disable_web_page_preview: false,
    }),
  });
  if (!res.ok) console.error(`[${nowStr()}] Telegram error: ${res.status}`);
}

function buildAlert(trade, trader) {
  const side     = (trade.side || "").toUpperCase();
  const emoji    = side === "BUY" ? "🟢" : "🔴";
  const size     = parseFloat(trade.size || 0);
  const price    = parseFloat(trade.price || 0);
  const title    = trade.title || trade.slug || "Unknown market";
  const outcome  = trade.outcome || "—";
  const slug     = trade.eventSlug || trade.slug || "";
  const link     = slug ? `${POLYMARKET_URL}/${slug}` : "https://polymarket.com";
  const ts       = trade.timestamp ? fmtTime(trade.timestamp) : "just now";

  return [
    `${emoji} <b>${side}</b>  ·  ${rankEmoji(trader.rank)} <b>${trader.name}</b>`,
    ``,
    `📋 <b>${title}</b>`,
    ``,
    `  Outcome:    <b>${outcome}</b>`,
    `  Price:      <b>${(price * 100).toFixed(1)}¢</b>`,
    `  Size:       <b>${fmtSize(size)}</b>`,
    `  Trader P&L: <b>${fmtMoney(trader.pnl)}</b> (all-time)`,
    ``,
    `🕐 ${ts}`,
    ``,
    `<a href="${link}">▶️ Open on Polymarket to mirror</a>`,
  ].join("\n");
}

async function sendStartup() {
  const lines = [
    `<b>📡 PolyTrack Bot started</b>`,
    ``,
    `Min trade size: <b>${fmtSize(MIN_SIZE)}</b>`,
    ``,
    `📌 <b>Fixed (All-Time)</b>`,
  ];
  for (const w of trackedWallets.filter(w => w.fixed)) {
    lines.push(`${rankEmoji(w.rank)} <b>${w.name}</b>  ${fmtMoney(w.pnl)}`);
    lines.push(`   <code>${w.wallet}</code>`);
  }
  lines.push(``, `📈 <b>Weekly Top 2</b>`);
  for (const w of trackedWallets.filter(w => !w.fixed)) {
    lines.push(`${rankEmoji(w.rank)} <b>${w.name}</b>  ${fmtMoney(w.pnl)} this week`);
    lines.push(`   <code>${w.wallet}</code>`);
  }
  lines.push(``, `Alerts will fire here when they trade.`);
  await sendTelegram(lines.join("\n"));
}

// ── Core ──────────────────────────────────────────────────────────────────────
async function refreshLeaderboard() {
  console.log(`[${nowStr()}] Refreshing leaderboard…`);

  // Always include swisstony — update his current PnL
  try {
    const alltime = await get(`${BASE}/v1/leaderboard?limit=10&orderBy=PNL&timePeriod=ALL`);
    const sw = alltime.find(t => t.proxyWallet === FIXED_WALLET.wallet);
    if (sw) FIXED_WALLET.pnl = parseFloat(sw.pnl || 0);
  } catch(e) {}

  // Get top 2 from weekly leaderboard, excluding swisstony
  const weekly = await get(`${BASE}/v1/leaderboard?limit=10&orderBy=PNL&timePeriod=WEEK`);
  const top2weekly = weekly
    .filter(t => t.proxyWallet !== FIXED_WALLET.wallet)
    .slice(0, 2)
    .map((t, i) => ({
      rank:   i + 2,
      name:   t.userName || t.proxyWallet?.slice(0, 8) + "…",
      wallet: t.proxyWallet || "",
      pnl:    parseFloat(t.pnl || 0),
      weekly: true,
    }));

  trackedWallets = [FIXED_WALLET, ...top2weekly];

  for (const w of trackedWallets) {
    const tag = w.fixed ? "ALL-TIME" : "WEEKLY";
    console.log(`  ${rankEmoji(w.rank)} ${w.name}  ${fmtMoney(w.pnl)}  [${tag}]`);
  }
  lastLeaderboardRefresh = Date.now();
}

async function seedSeen() {
  console.log(`[${nowStr()}] Seeding trade history (won't replay old trades)…`);
  for (const trader of trackedWallets) {
    if (!trader.wallet) continue;
    try {
      const trades = await get(`${BASE}/trades?user=${trader.wallet}&limit=100&takerOnly=false`);
      (Array.isArray(trades) ? trades : []).forEach(t => { if (t.transactionHash) seenTx.add(t.transactionHash); });
      console.log(`  Seeded ${trades.length || 0} trades for ${trader.name}`);
    } catch (e) {
      console.error(`  Seed failed for ${trader.name}: ${e.message}`);
    }
  }
}

async function fetchWinLoss(trader) {
  try {
    let positions = [];

    // Try redeemed positions first
    try {
      const d = await get(`${BASE}/positions?user=${trader.wallet}&limit=500&redeemed=true`);
      positions = Array.isArray(d) ? d : (d.positions || d.data || []);
    } catch(e) {}

    // Fallback: get all positions and filter resolved ones client-side
    if (positions.length === 0) {
      try {
        const d = await get(`${BASE}/positions?user=${trader.wallet}&limit=500`);
        const all = Array.isArray(d) ? d : (d.positions || d.data || []);
        positions = all.filter(p => {
          const redeemed = p.redeemed === true || p.redeemed === 1;
          const resolved = parseFloat(p.currentValue ?? p.currentPrice ?? -1);
          const zeroed   = parseFloat(p.size ?? 1) === 0;
          return redeemed || resolved === 0 || resolved >= 0.99 || zeroed;
        });
      } catch(e) {}
    }

    // Last resort: use trades endpoint and derive from cashPnl field
    if (positions.length === 0) {
      const d = await get(`${BASE}/trades?user=${trader.wallet}&limit=500&takerOnly=false`);
      const trades = Array.isArray(d) ? d : [];
      // Group by conditionId to get per-market result
      const markets = {};
      for (const t of trades) {
        const id = t.conditionId;
        if (!id) continue;
        if (!markets[id]) markets[id] = { pnl: 0, size: 0 };
        const pnl  = parseFloat(t.cashPnl ?? 0);
        const size = parseFloat(t.size ?? 0);
        markets[id].pnl  += pnl;
        markets[id].size += size;
      }
      positions = Object.values(markets).filter(m => m.pnl !== 0);
    }

    let wins = 0, losses = 0, totalPnl = 0, halfDollarPnl = 0;

    for (const p of positions) {
      const pnl  = parseFloat(p.cashPnl ?? p.pnl ?? 0);
      const size = parseFloat(p.size ?? p.initialValue ?? 1);
      totalPnl += pnl;

      // Scale to $0.50: if they risked $size and made $pnl, $0.50 would make pnl*(0.50/size)
      if (size > 0) halfDollarPnl += pnl * (0.50 / size);

      if (pnl > 0) wins++;
      else if (pnl < 0) losses++;
    }

    const total   = wins + losses;
    const winRate = total > 0 ? Math.round((wins / total) * 100) : 0;

    return { wins, losses, total, winRate, totalPnl, halfDollarPnl, tracked: positions.length };
  } catch (e) {
    return null;
  }
}

async function sendWinLossReport() {
  const lines = [`<b>📊 Win/Loss Report</b>`, `<i>Settled markets only</i>`, ``];

  for (const trader of trackedWallets) {
    const r = await fetchWinLoss(trader);
    if (!r || r.total === 0) {
      lines.push(`${rankEmoji(trader.rank)} <b>${trader.name}</b> — no settled data yet`);
      lines.push(``);
      continue;
    }

    const sign = r.halfDollarPnl >= 0 ? "+" : "";
    const pnlSign = r.totalPnl >= 0 ? "+" : "";

    lines.push(`${rankEmoji(trader.rank)} <b>${trader.name}</b>`);
    lines.push(`  Trades tracked:       <b>${r.tracked}</b>`);
    lines.push(`  Resolved:             <b>${r.total}</b>`);
    lines.push(`  Won: <b>${r.wins}</b>  ·  Lost: <b>${r.losses}</b>`);
    lines.push(`  Win rate:             <b>${r.winRate}%</b>`);
    lines.push(`  Your P&L ($0.50/trade): <b>${sign}$${Math.abs(r.halfDollarPnl).toFixed(2)}</b>`);
    lines.push(`  Their P&L:            <b>${pnlSign}${fmtMoney(r.totalPnl).replace('+','').replace('-','')}</b>`);
    lines.push(``);
  }

  await sendTelegram(lines.join("\n"));
}

async function pollTrades() {
  for (const trader of trackedWallets) {
    if (!trader.wallet) continue;
    let trades;
    try { trades = await fetchTrades(trader.wallet); }
    catch (e) { console.error(`[${nowStr()}] Trades error (${trader.name}): ${e.message}`); continue; }

    for (const trade of trades) {
      const tx   = trade.transactionHash || "";
      const size = parseFloat(trade.size || 0);
      if (!tx || seenTx.has(tx)) continue;
      seenTx.add(tx);
      if (size < MIN_SIZE) continue;

      console.log(`[${nowStr()}] 🔔 NEW — ${trader.name}: ${trade.side} ${fmtSize(size)}`);
      await sendTelegram(buildAlert(trade, trader));
      await new Promise(r => setTimeout(r, 500));
    }
  }
}

async function main() {
  console.log("════════════════════════════");
  console.log("  PolyTrack Bot");
  console.log("════════════════════════════\n");

  if (!TELEGRAM_TOKEN) console.warn("[WARN] TELEGRAM_TOKEN not set — alerts print here instead\n");

  await refreshLeaderboard();
  await seedSeen();
  await sendStartup();

  // Send win/loss report on startup
  await sendWinLossReport();

  console.log(`\n[${nowStr()}] Running. Checks every ${POLL_INTERVAL/1000}s. Ctrl+C to stop.\n`);

  setInterval(async () => {
    try {
      if (Date.now() - lastLeaderboardRefresh > LEADERBOARD_TTL) await refreshLeaderboard();
      await pollTrades();
    } catch (e) {
      console.error(`[${nowStr()}] Loop error: ${e.message}`);
    }
  }, POLL_INTERVAL);

  // Send win/loss report every 24 hours
  setInterval(async () => {
    await sendWinLossReport();
  }, 24 * 60 * 60 * 1000);

  // keep process alive
  await pollTrades();
}

main().catch(e => { console.error("Fatal:", e.message); process.exit(1); });

File: bot.py
"""
Polymarket Top Trader Tracker — Telegram Bot
Polls the top 3 leaderboard traders and sends alerts when they make new trades.
"""

import os
import time
import json
import requests
from datetime import datetime, timezone

# ── Config ──────────────────────────────────────────────────────────────────
TELEGRAM_TOKEN  = os.environ.get("TELEGRAM_TOKEN", "")   # from @BotFather
TELEGRAM_CHAT   = os.environ.get("TELEGRAM_CHAT", "")    # your personal chat ID
POLL_INTERVAL   = int(os.environ.get("POLL_INTERVAL", "120"))  # seconds
LEADERBOARD_PERIOD = os.environ.get("LEADERBOARD_PERIOD", "ALL")  # DAY/WEEK/MONTH/ALL
MIN_TRADE_SIZE  = float(os.environ.get("MIN_TRADE_SIZE", "0.50"))

BASE = "https://data-api.polymarket.com"
POLYMARKET_URL = "https://polymarket.com/event"

# ── State ────────────────────────────────────────────────────────────────────
seen_tx_hashes: set[str] = set()
tracked_wallets: list[dict] = []   # [{wallet, name, rank, pnl}]
last_leaderboard_refresh = 0
LEADERBOARD_TTL = 3600  # refresh top-3 every hour


# ── Helpers ──────────────────────────────────────────────────────────────────
def fmt_money(n: float, sign: bool = True) -> str:
    n = float(n or 0)
    prefix = ("+" if n >= 0 else "-") if sign else ""
    abs_n = abs(n)
    if abs_n >= 1_000_000:
        return f"{prefix}${abs_n/1e6:.2f}M"
    if abs_n >= 1_000:
        return f"{prefix}${abs_n/1e3:.1f}K"
    return f"{prefix}${abs_n:.2f}"


def fmt_size(n: float) -> str:
    n = float(n or 0)
    if n >= 1_000_000:
        return f"${n/1e6:.2f}M"
    if n >= 1_000:
        return f"${n/1e3:.1f}K"
    return f"${n:.2f}"


def rank_emoji(rank: int) -> str:
    return {1: "🥇", 2: "🥈", 3: "🥉"}.get(rank, f"#{rank}")


def side_emoji(side: str) -> str:
    return "🟢" if side.upper() == "BUY" else "🔴"


def price_to_cents(price: float) -> str:
    return f"{price * 100:.1f}¢"


def ts_to_str(ts: int) -> str:
    dt = datetime.fromtimestamp(ts, tz=timezone.utc)
    return dt.strftime("%b %d %H:%M UTC")


# ── API calls ────────────────────────────────────────────────────────────────
def fetch_leaderboard() -> list[dict]:
    r = requests.get(
        f"{BASE}/v1/leaderboard",
        params={"limit": 3, "orderBy": "PNL", "timePeriod": LEADERBOARD_PERIOD},
        timeout=15,
    )
    r.raise_for_status()
    return r.json()


def fetch_trades(wallet: str, limit: int = 50) -> list[dict]:
    r = requests.get(
        f"{BASE}/trades",
        params={"user": wallet, "limit": limit, "takerOnly": "false"},
        timeout=15,
    )
    r.raise_for_status()
    data = r.json()
    return data if isinstance(data, list) else []


# ── Telegram ─────────────────────────────────────────────────────────────────
def send_telegram(text: str) -> bool:
    if not TELEGRAM_TOKEN or not TELEGRAM_CHAT:
        print("[WARN] Telegram not configured — printing to console instead")
        print(text)
        return True
    url = f"https://api.telegram.org/bot{TELEGRAM_TOKEN}/sendMessage"
    payload = {
        "chat_id": TELEGRAM_CHAT,
        "text": text,
        "parse_mode": "HTML",
        "disable_web_page_preview": False,
    }
    try:
        r = requests.post(url, json=payload, timeout=10)
        r.raise_for_status()
        return True
    except Exception as e:
        print(f"[ERROR] Telegram send failed: {e}")
        return False


def build_trade_message(trade: dict, trader: dict) -> str:
    side = trade.get("side", "").upper()
    size = float(trade.get("size", 0))
    price = float(trade.get("price", 0))
    title = trade.get("title") or trade.get("slug") or "Unknown market"
    outcome = trade.get("outcome") or "—"
    event_slug = trade.get("eventSlug") or trade.get("slug") or ""
    ts = trade.get("timestamp")

    market_url = f"{POLYMARKET_URL}/{event_slug}" if event_slug else "https://polymarket.com"

    # implied probability move hint
    prob_hint = ""
    if price > 0:
        if price < 0.15:
            prob_hint = "  ⚠️ Low prob bet"
        elif price > 0.85:
            prob_hint = "  ✅ High confidence"

    lines = [
        f"{side_emoji(side)} <b>{side}</b>  ·  {rank_emoji(trader['rank'])} <b>{trader['name']}</b>",
        f"",
        f"📋 <b>{title}</b>",
        f"",
        f"  Outcome:  <b>{outcome}</b>",
        f"  Price:    <b>{price_to_cents(price)}</b>{prob_hint}",
        f"  Size:     <b>{fmt_size(size)}</b>",
        f"  Trader PnL:  <b>{fmt_money(trader['pnl'])}</b>  (all-time)",
        f"",
        f"🕐 {ts_to_str(ts) if ts else 'just now'}",
        f"",
        f"<a href=\"{market_url}\">▶️ Open on Polymarket to mirror</a>",
    ]
    return "\n".join(lines)


def send_startup_message(wallets: list[dict]) -> None:
    lines = ["<b>📡 PolyTrack Bot started</b>", ""]
    lines.append(f"Tracking top {len(wallets)} traders — period: <b>{LEADERBOARD_PERIOD}</b>")
    lines.append(f"Min trade size: <b>{fmt_size(MIN_TRADE_SIZE)}</b>")
    lines.append(f"Poll interval: <b>{POLL_INTERVAL}s</b>")
    lines.append("")
    for w in wallets:
        lines.append(
            f"{rank_emoji(w['rank'])} <b>{w['name']}</b>  {fmt_money(w['pnl'])}"
        )
        lines.append(f"   <code>{w['wallet']}</code>")
    lines.append("")
    lines.append("Alerts will fire here when they trade.")
    send_telegram("\n".join(lines))


# ── Core loop ────────────────────────────────────────────────────────────────
def refresh_leaderboard() -> None:
    global tracked_wallets, last_leaderboard_refresh
    print(f"[{ts_now()}] Refreshing leaderboard…")
    try:
        leaders = fetch_leaderboard()
        tracked_wallets = []
        for i, trader in enumerate(leaders[:3], start=1):
            wallet = trader.get("proxyWallet") or ""
            name = trader.get("userName") or wallet[:8] + "…"
            pnl = float(trader.get("pnl") or 0)
            tracked_wallets.append({"rank": i, "name": name, "wallet": wallet, "pnl": pnl})
            print(f"  {rank_emoji(i)} {name}  {fmt_money(pnl)}  {wallet}")
        last_leaderboard_refresh = time.time()
    except Exception as e:
        print(f"[ERROR] Leaderboard fetch failed: {e}")


def poll_trades() -> None:
    for trader in tracked_wallets:
        wallet = trader["wallet"]
        if not wallet:
            continue
        try:
            trades = fetch_trades(wallet, limit=20)
        except Exception as e:
            print(f"[ERROR] Trades for {wallet[:8]}…: {e}")
            continue

        new_count = 0
        for trade in trades:
            tx = trade.get("transactionHash") or ""
            if not tx or tx in seen_tx_hashes:
                continue

            size = float(trade.get("size", 0))
            if size < MIN_TRADE_SIZE:
                seen_tx_hashes.add(tx)
                continue

            seen_tx_hashes.add(tx)
            new_count += 1
            msg = build_trade_message(trade, trader)
            print(f"[{ts_now()}] NEW TRADE from {trader['name']}: {trade.get('side')} {fmt_size(size)}")
            send_telegram(msg)
            time.sleep(0.5)  # don't spam Telegram

        if new_count:
            print(f"  → Sent {new_count} alert(s) for {trader['name']}")


def seed_seen_hashes() -> None:
    """On startup, load existing trades so we don't replay history."""
    print(f"[{ts_now()}] Seeding known trades to suppress history…")
    for trader in tracked_wallets:
        wallet = trader["wallet"]
        if not wallet:
            continue
        try:
            trades = fetch_trades(wallet, limit=100)
            for t in trades:
                tx = t.get("transactionHash")
                if tx:
                    seen_tx_hashes.add(tx)
            print(f"  Seeded {len(trades)} trades for {trader['name']}")
        except Exception as e:
            print(f"[ERROR] Seed failed for {wallet[:8]}…: {e}")


def ts_now() -> str:
    return datetime.now(tz=timezone.utc).strftime("%H:%M:%S")


def main() -> None:
    print("=" * 50)
    print("  PolyTrack Bot")
    print("=" * 50)

    if not TELEGRAM_TOKEN:
        print("\n[WARN] TELEGRAM_TOKEN not set. Messages will print to console.\n")

    refresh_leaderboard()

    if not tracked_wallets:
        print("[ERROR] No wallets to track. Check API connectivity.")
        return

    seed_seen_hashes()
    send_startup_message(tracked_wallets)

    print(f"\n[{ts_now()}] Polling every {POLL_INTERVAL}s. Ctrl+C to stop.\n")

    while True:
        try:
            # Refresh leaderboard hourly
            if time.time() - last_leaderboard_refresh > LEADERBOARD_TTL:
                refresh_leaderboard()

            poll_trades()
            time.sleep(POLL_INTERVAL)

        except KeyboardInterrupt:
            print("\nStopped.")
            break
        except Exception as e:
            print(f"[ERROR] Unexpected: {e}")
            time.sleep(30)


if __name__ == "__main__":
    main()
